src/universal/components/SignInPage/SignInPage.js

```diff
@@ -0,0 +1,222 @@
+/**
+ * The sign-in page.  Hosts 3rd part signin, email/password signin, and
+ * also functions as the callback handler for the Auth0 OIDC response.
+ *
+ * @flow
+ */
+import type {Dispatch} from 'redux';
+import type {RouterHistory, Location} from 'react-router-dom';
+
+import React, {Component} from 'react';
+import Loadable from 'react-loadable';
+import {connect} from 'react-redux';
+import {Link, withRouter} from 'react-router-dom';
+
+import signinAndUpdateToken from 'universal/components/Auth0ShowLock/signinAndUpdateToken';
+import LoadableLoading from 'universal/components/LoadableLoading';
+import LoadingView from 'universal/components/LoadingView/LoadingView';
+import withAtmosphere from 'universal/decorators/withAtmosphere/withAtmosphere';
+
+import Header from './Header';
+import SignIn from './SignIn';
+import chunkArray from 'universal/utils/chunkArray';
+
+type Credentials = {
+  email: string,
+  password: string
+};
+
+type ParsedAuthResponse = {
+  idToken: string,
+  idTokenPayload: Object
+};
+
+type Props = {
+  atmosphere: Object,
+  dispatch: Dispatch<*>,
+  hasSession: boolean,
+  history: RouterHistory,
+  location: Location,
+  webAuth: Object
+};
+
+type State = {
+  loggingIn: boolean,
+  error: ?Error
+};
+
+const containerStyles = {
+  display: 'flex',
+  flexDirection: 'column'
+};
+
+const loadingErrorStyles = ({
+  display: 'flex',
+  flexDirection: 'column',
+  alignItems: 'center',
+  height: '100%'
+});
+
+// A naive implementation taking a querystring-formatted string (beginning in '?'),
+// and returning an object of the key/value pairs.  Will not work for null-valued
+// keys, but works for our own `?redirectTo=/path/to/page` contract.
+const parseSearch = (search: string): Object => {
```

Comment:
this makes me a little nervous. i think we should at least use `decodeURIComponent` since browsers will automatically encode the string & that could make it barf. The _right_ way to do it would be using the browser native url search params: https://davidwalsh.name/query-string-javascript

Reply:
Well isn't that handy!